File: lib/allure_report_publisher/lib/uploaders/s3.rb
require "aws-sdk-s3"
require "mini_mime"

module Publisher
  module Uploaders
    # Report upload to AWS S3 bucket
    #
    class S3 < Uploader
      # Report url
      #
      # @return [String]
      def report_url
        @report_url ||= url(full_prefix)
      end

      private

      # S3 client
      #
      # @return [Aws::S3::Client]
      def client
        @client ||= Aws::S3::Client.new(client_args)
      rescue Aws::Sigv4::Errors::MissingCredentialsError
        raise(<<~MSG.strip)
          missing aws credentials, provide credentials with one of the following options:
            - AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables
            - ~/.aws/credentials file
        MSG
      end

      def client_args
        @client_args ||= {
          region: ENV["AWS_REGION"] || "us-east-1",
          force_path_style: path_style?,
          endpoint: aws_endpoint
        }.compact
      end

      # Latest report url
      #
      # @return [String]
      def latest_report_url
        @latest_report_url ||= url(prefix)
      end

      # Custom aws endpoint
      #
      # @return [<String, nil>]
      def aws_endpoint
        @aws_endpoint ||= ENV["AWS_ENDPOINT"]
      end

      # Force aws path style
      #
      # @return [Boolean]
      def path_style?
        ENV["AWS_FORCE_PATH_STYLE"] == "true"
      end

      # Add allure history
      #
      # @return [void]
      def download_history
        log_debug("Downloading previous run history")
        HISTORY.each do |file_name|
          file_path = path(common_info_path, "history", file_name)
          client.get_object(
            response_target: file_path,
            key: key(prefix, "history", file_name),
            bucket: bucket_name
          )
          log_debug("Downloaded '#{file_name}' as '#{file_path}'")
        end
      rescue Aws::S3::Errors::NoSuchKey
        raise(HistoryNotFoundError, "Allure history from previous runs not found!")
      end

      # Upload allure history
      #
      # @return [void]
      def upload_history
        log_debug("Uploading report history")
        upload_to_s3(report_files.select { |file| file.fnmatch?("*/history/*") }, prefix)
      end

      # Upload allure report
      #
      # @return [void]
      def upload_report
        log_debug("Uploading report files")
        upload_to_s3(report_files, full_prefix)
      end

      # Upload copy of latest run
      #
      # @return [void]
      def upload_latest_copy
        log_debug("Copying report as latest")

        # it's not possible to overwrite whole directory so we clean out unique data files from last run
        log_debug("Cleaning data files")
        client.list_objects_v2(bucket: bucket_name, prefix: key(prefix, "data")).each do |resp|
          data_files = resp.contents.map { |obj| { key: obj.key } }
          next if data_files.empty?

          client.delete_objects({ bucket: bucket_name, delete: { objects: data_files } })
        end

        log_debug("Copying report files")
        args = report_files.map do |file|
          {
            bucket: bucket_name,
            copy_source: "/#{bucket_name}/#{key(full_prefix, file.relative_path_from(report_path))}",
            key: key(prefix, file.relative_path_from(report_path)),
            metadata_directive: "REPLACE",
            content_type: content_type(file),
            cache_control: "max-age=60"
          }
        end
        Parallel.each(args, in_threads: parallel) { |obj| client.copy_object(obj) }
        log_debug("Finished latest report copy successfully")
      end

      # Upload files to s3
      #
      # @param [Array<Pathname>] files
      # @param [String] key_prefix
      # @return [Array<Hash>]
      def upload_to_s3(files, key_prefix)
        args = files.map do |file|
          {
            body: file.to_s,
            bucket: bucket_name,
            key: key(key_prefix, file.relative_path_from(report_path)),
            content_type: content_type(file),
            cache_control: "max-age=3600"
          }
        end

        log_debug("Uploading '#{args.size}' files in '#{parallel}' threads")
        Parallel.each(args, in_threads: parallel) do |obj|
          client.put_object(obj.tap { |o| o[:body] = File.read(o[:body]) })
        end
        log_debug("Finished upload successfully")
      end

      # Fabricate key for s3 object
      #
      # @param [String] *args
      # @return [String]
      def key(*args)
        args.compact.join("/")
      end

      # Report url
      #
      # @param [String] path_prefix
      # @return [String]
      def url(path_prefix)
        custom_base = if aws_endpoint
                        path_style? ? "#{aws_endpoint}/#{bucket_name}" : aws_endpoint
                      end

        [
          base_url || custom_base || "http://#{bucket_name}.s3.amazonaws.com",
          path_prefix,
          "index.html"
        ].compact.join("/")
      end

      # Get file content type
      #
      # @param [Pathname] file
      # @return [String]
      def content_type(file)
        c_type = MiniMime.lookup_by_filename(file)&.content_type
        return c_type if c_type

        log_debug("Content type for '#{file}' not found, using default 'application/octet-stream'")
        "application/octet-stream"
      end
    end
  end
end

File: src/lib/uploader/s3.ts
import {
  CopyObjectCommand,
  GetObjectCommand,
  NoSuchKey,
  PutObjectCommand,
  S3Client,
  waitUntilObjectExists,
} from '@aws-sdk/client-s3'
import {lookup} from 'mime-types'
import {readFileSync, writeFileSync} from 'node:fs'
import path from 'node:path'
import pAll from 'p-all'

import {config} from '../../utils/config.js'
import {logger} from '../../utils/logger.js'
import {BaseUploader} from './base.js'

export class S3Uploader extends BaseUploader {
  private _reportUrlBase: string | undefined
  private readonly s3Client: S3Client = new S3Client({
    endpoint: this.awsEndpoint,
    forcePathStyle: this.forcePathStyle,
    region: process.env.AWS_REGION || 'us-east-1',
  })

  private get awsEndpoint() {
    return process.env.AWS_ENDPOINT
  }

  private get forcePathStyle() {
    return process.env.AWS_FORCE_PATH_STYLE === 'true'
  }

  private get region() {
    return process.env.AWS_REGION || 'us-east-1'
  }

  protected reportUrlBase() {
    if (this._reportUrlBase) return this._reportUrlBase

    let base = `https://${this.bucketName}.s3.${this.region}.amazonaws.com`
    if (this.awsEndpoint) {
      base = this.forcePathStyle ? `${this.awsEndpoint}/${this.bucketName}` : this.awsEndpoint
    }

    this._reportUrlBase = this.prefix ? `${base}/${this.prefix}` : base
    return this._reportUrlBase
  }

  public async downloadHistory() {
    await super.downloadHistory()

    const key = this.key(this.historyFileName())
    logger.debug(`Downloading history file from s3://${this.bucketName}/${key}`)

    try {
      const response = await this.s3Client.send(
        new GetObjectCommand({
          Bucket: this.bucketName,
          Key: key,
        }),
      )

      if (!response.Body) {
        throw new Error('Received empty response body when downloading history file')
      }

      const content = await response.Body.transformToString()
      if (content.trim().length === 0) throw new Error('Received empty response body when downloading history file')

      writeFileSync(this.historyPath, content, 'utf8')
      logger.debug(`History file downloaded successfully to ${this.historyPath}`)
    } catch (error) {
      throw error instanceof NoSuchKey ? new Error(`History file not found!`) : error
    }
  }

  protected async uploadHistory() {
    logger.debug('Uploading history file')
    await this.uploadFile({
      filePath: this.historyPath,
      key: this.key(this.historyFileName()),
    })
  }

  protected async uploadReport() {
    logger.debug(`Uploading report files with concurrency: ${this.parallel}`)
    const uploads = (await this.getReportFiles()).map((filePath) => {
      const key = this.key(this.runId, path.relative(this.reportPath, filePath))
      return () => this.uploadFile({filePath, key})
    })

    await pAll(uploads, {concurrency: config.parallel})
  }

  protected async createLatestCopy() {
    logger.debug(`Creating latest report copy with concurrency: ${this.parallel}`)
    const copies = (await this.getReportFiles()).map((filePath) => {
      const sourceKey = this.key(this.runId, path.relative(this.reportPath, filePath))
      const destinationKey = this.key('latest', path.relative(this.reportPath, filePath))
      return () => this.copyFile({sourceKey, destinationKey})
    })

    await pAll(copies, {concurrency: config.parallel})
  }

  private async uploadFile(opts: {cacheControl?: string; filePath: string; key: string}) {
    const content = readFileSync(opts.filePath)
    const contentType = lookup(opts.filePath) || 'application/octet-stream'

    logger.debug(`- uploading '${opts.filePath}' to 's3://${this.bucketName}/${opts.key}' (type: ${contentType})`)
    return this.s3Client.send(
      new PutObjectCommand({
        Body: content,
        Bucket: this.bucketName,
        Key: opts.key,
        CacheControl: opts.cacheControl || 'max-age=3600',
        ContentType: contentType,
      }),
    )
  }

  private async copyFile(opts: {sourceKey: string; destinationKey: string}) {
    const source = `${this.bucketName}/${opts.sourceKey}`

    logger.debug(`- copying 's3://${source}' to 's3://${this.bucketName}/${opts.destinationKey}'`)
    await this.s3Client.send(
      new CopyObjectCommand({
        CopySource: source,
        Bucket: this.bucketName,
        Key: opts.destinationKey,
      }),
    )
    await waitUntilObjectExists(
      {client: this.s3Client, maxWaitTime: 60},
      {Bucket: this.bucketName, Key: opts.destinationKey},
    )
  }
}
